### PR TITLE
add fault tolerance to train

### PR DESCRIPTION
**Pull request by cainiaogoroad**

### CODE REVIEW

File: mlora/executor/task/train_task.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 from mlora.config import TaskConfig
 from mlora.model.args import LinearInfo, Tokens, Masks, MLoRADataConfig
 from mlora.model.tokenizer import Tokenizer
-
+import mlora.utils
```
Comment (yezhengmao1) [inline mlora/executor/task/train_task.py:4]: why train task module needs the utils, the mlora.utils only export to user.

File: mlora/utils/deletefile.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import os
+import shutil
+import os.path
+
+def delete_files_in_folder(folder_path):
+    try:
+        # 遍历目录中的每个项目
```
Comment (yezhengmao1) [inline mlora/utils/deletefile.py:7]: comment be english..

**Review by yezhengmao1 — CHANGES_REQUESTED** [review]

File: mlora/executor/task/train_task.py (hunk 1 above)
Comment (cainiaogoroad) [reply mlora/executor/task/train_task.py:4]: because i need the function delete_files_in_folder(folder_path) of  deletefile.py in the utils

File: mlora/utils/deletefile.py (hunk 2 above)
Comment (cainiaogoroad) [reply mlora/utils/deletefile.py:7]: ok,i will modify later

Comment (cainiaogoroad) [reply mlora/utils/deletefile.py]: finished

File: mlora/utils/deletefile.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import os
+import shutil
+import os.path
+
+def delete_files_in_folder(folder_path):
+    try:
+        # Iterate through each file in the directory
+        for item in os.listdir(folder_path):
+            item_path = os.path.join(folder_path, item)
+            if os.path.isfile(item_path):  # If it's a file, delete it
```
Comment (yezhengmao1) [inline mlora/utils/deletefile.py:10]: i think we do not need those comment, just known what you want to do by reading those code.

File: mlora/executor/task/train_task.py [hunk 4]
```diff
@@ -96,13 +97,21 @@ def _save(self, dir_suffix: str = "", additional_info: Dict[str, str] = {}):
         adapter_config["base_model_name_or_path"] = self.llm_name_
         adapter_config = {**adapter_config, **additional_info}
         adapter_config = {**adapter_config, **self.config_.adapter_.export()}
+        logging.info(f"save optimizer weight")
```
Comment (yezhengmao1) [inline mlora/executor/task/train_task.py:100]: the log maybe confuse, which optimizer be saved.

File: mlora/executor/context/context.py [hunk 5]
```diff
@@ -11,16 +11,16 @@ class TaskContext(metaclass=ABCMeta):
     type_: str = ""
     name_: str = ""
     path_: str = ""
-
+    now_step_: int = 0
     device_: str = ""
-
+    last_epoch: int = -1
```
Comment (yezhengmao1) [inline mlora/executor/context/context.py:16]: what's the meaning of the last epoch?

**Review by yezhengmao1 — CHANGES_REQUESTED** [review]

File: mlora/utils/deletefile.py (hunk 3 above)
Comment (cainiaogoroad) [reply mlora/utils/deletefile.py:10]: ok

File: mlora/executor/task/train_task.py (hunk 4 above)
Comment (cainiaogoroad) [reply mlora/executor/task/train_task.py]: ok,i will modify later

File: mlora/executor/context/context.py (hunk 5 above)
Comment (cainiaogoroad) [reply mlora/executor/context/context.py:16]: Make the learning rate to be same as the state before fault  which we saved

Comment (cainiaogoroad) [reply mlora/executor/context/context.py:16]: torch.optim.lr_scheduler.CosineAnnealingLR(optimizer,T_max,eta_min=0,last_epoch=-1) can use the parameter "last_epoch" to ensure the learning rate to be same as the state before fault which we saved

File: mlora/config/lr_scheduler.py [hunk 6]
```diff
@@ -14,6 +14,7 @@ class LRSchedulerConfig(DictConfig):
     def __init__(self, config: Dict[str, str]) -> None:
         super().__init__(config)
         self.init(self.__params_map, config)
+        self.last_epoch = -1
```
Comment (yezhengmao1) [inline mlora/config/lr_scheduler.py:17]: why we need add new item named last_epoch?

File: mlora/executor/task/task.py [hunk 7]
```diff
@@ -4,7 +4,7 @@
 from mlora.model.args import LinearInfo, Tokens, Masks, MLoRADataConfig
 from mlora.model.tokenizer import Tokenizer
 from mlora.executor.context import TaskContext, TRAINCONTEXT_CLASS
-
+import ipdb
```
Comment (yezhengmao1) [inline mlora/executor/task/task.py:7]: do not import ipdb!

**Review by yezhengmao1 — CHANGES_REQUESTED** [review]

File: mlora/executor/task/task.py (hunk 7 above)
Comment (cainiaogoroad) [reply mlora/executor/task/task.py:7]: ok

File: mlora/config/lr_scheduler.py (hunk 6 above)
Comment (cainiaogoroad) [reply mlora/config/lr_scheduler.py:17]: because the epoch is the only parameter to decide learning rate in function  torch.optim.lr_scheduler.CosineAnnealingLR which we use ,so we need this new item to make sure the learning rate to be same as the state we saved before.
